algorithms/kernel/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i

```diff
@@ -71,289 +92,249 @@ Status KNNClassificationPredictKernelUCAPI<algorithmFpType>::compute(const Numer
+            DAAL_CHECK_STATUS_VAR(points->releaseBlockOfRows(dataRows));
         }
-        RadixSort::sort(selectResult.indices, sorted, radix_buffer, curProbeBlockSize, nK, nK, &st);
+        // force merging of remainig partial data
+        mergePartialSelection(true, partialDistances, partialLabels, partCount, selectionNumberOfChunks, curQueryRows.number, nK, selector,
+                              selectResult, &st);
         DAAL_CHECK_STATUS_VAR(st);
-        {
-            DAAL_ITTNOTIFY_SCOPED_TASK(compute.winners);
-            computeWinners(context, find_winners, sorted, classes, curProbeBlockSize, nK, &st);
-            DAAL_CHECK_STATUS_VAR(st);
-        }
-        {
-            DAAL_ITTNOTIFY_SCOPED_TASK(compute.copy_results);
-            BlockDescriptor<algorithmFpType> resultBlock;
-            DAAL_CHECK_STATUS_VAR(y->getBlockOfRows(0, nProbeRows, writeOnly, resultBlock));
-            auto classesRows = classes.template get<int>().toHost(ReadWriteMode::readOnly);
-            auto res         = resultBlock.getBuffer().toHost(ReadWriteMode::writeOnly);
-            for (int i = 0; i < curProbeBlockSize; i++)
-            {
-                res.get()[pfirst + i] = classesRows.get()[i];
-            }
-            DAAL_CHECK_STATUS_VAR(y->releaseBlockOfRows(resultBlock));
-        }
+        // sort labels of closest neighbors
+        RadixSort::sort(selectResult.indices, sorted, radix_buffer, curQueryRows.number, nK, nK, &st);
+        DAAL_CHECK_STATUS_VAR(st);
+        BlockDescriptor<int> resultBlock;
+        DAAL_CHECK_STATUS_VAR(y->getBlockOfRows(curQueryRows.first, curQueryRows.number, writeOnly, resultBlock));
+        auto classes = UniversalBuffer(resultBlock.getBuffer());
+        // search for maximum occurrence label
+        computeWinners(context, sorted, classes, curQueryRows.number, nK, &st);
+        DAAL_CHECK_STATUS_VAR(st);
+        DAAL_CHECK_STATUS_VAR(y->releaseBlockOfRows(resultBlock));
+        DAAL_CHECK_STATUS_VAR(ntData->releaseBlockOfRows(queryRows));
     }
-    DAAL_CHECK_STATUS_VAR(ntData->releaseBlockOfRows(probeRows));
-    DAAL_CHECK_STATUS_VAR(points->releaseBlockOfRows(dataRows));
-    DAAL_CHECK_STATUS_VAR(labels->releaseBlockOfRows(labelRows));
     return st;
 }
-
 template <typename algorithmFpType>
-void KNNClassificationPredictKernelUCAPI<algorithmFpType>::initCategories(ExecutionContextIface & context, const KernelPtr & kernel_init_categories,
-                                                                          const Buffer<int> & labels, UniversalBuffer & categories,
-                                                                          uint32_t probeBlockSize, uint32_t dataBlockSize, uint32_t offset,
-                                                                          Status * st)
+void KNNClassificationPredictKernelUCAPI<algorithmFpType>::copyPartialSelections(ExecutionContextIface & context, const UniversalBuffer & distances,
```

Comment:
There is no parameter named `partialSelections` in this function. That is confusing. Probably, we should name it `copyPartialDistancesAndLabels`? Also, the kernel has the name "gather_partial_selection" - is it planned to give different names for kernel and driver for this function?